File: mobile/ios/OnlyCoffee/Models/PersonalizedOffer.swift
import Foundation

/// Personalized offer model
struct PersonalizedOffer: Codable, Identifiable {
    let id: String
    let title: String
    let description: String
    let offerType: OfferType
    let offerValue: Double
    let imageUrl: String?
    let validFrom: Date?
    let validUntil: Date?
    let minPurchaseAmount: Double?
    let maxDiscountAmount: Double?
    let targetSegments: [String]?
    let priority: Int
    let source: OfferSource

    // Personalization context
    let personalizedFor: String? // User segment
    let confidenceScore: Double?
    let reason: String? // Why this offer was selected

    // CTA
    let ctaText: String
    let deepLink: String?

    var isValid: Bool {
        let now = Date()
        if let validFrom = validFrom, now < validFrom {
            return false
        }
        if let validUntil = validUntil, now > validUntil {
            return false
        }
        return true
    }

    var formattedOfferValue: String {
        switch offerType {
        case .percentageDiscount:
            return "\(Int(offerValue))% OFF"
        case .fixedDiscount:
            return "$\(String(format: "%.2f", offerValue)) OFF"
        case .freeItem:
            return "FREE Item"
        case .buyXGetY:
            return "Buy X Get Y"
        case .pointsBonus:
            return "\(Int(offerValue))x Points"
        }
    }

    var validityText: String {
        guard let validUntil = validUntil else {
            return "No expiration"
        }

        let calendar = Calendar.current
        let now = Date()

        if calendar.isDateInToday(validUntil) {
            return "Expires today"
        } else if calendar.isDateInTomorrow(validUntil) {
            return "Expires tomorrow"
        } else {
            let days = calendar.dateComponents([.day], from: now, to: validUntil).day ?? 0
            if days < 7 {
                return "Expires in \(days) days"
            } else {
                let formatter = DateFormatter()
                formatter.dateStyle = .short
                return "Expires \(formatter.string(from: validUntil))"
            }
        }
    }
}

enum OfferType: String, Codable {
    case percentageDiscount = "percentage_discount"
    case fixedDiscount = "fixed_discount"
    case freeItem = "free_item"
    case buyXGetY = "buy_x_get_y"
    case pointsBonus = "points_bonus"
}

enum OfferSource: String, Codable {
    case aiGenerated = "ai_generated"
    case manual = "manual"
    case geofence = "geofence"
    case bandit = "bandit"
    case trigger = "trigger"
}

/// Response from personalized offers API
struct PersonalizedOffersResponse: Codable {
    let success: Bool
    let offers: [PersonalizedOffer]
    let count: Int
    let userSegment: String?
    let churnRisk: Double?
}

File: android/app/src/main/java/com/onlycoffee/app/data/model/PersonalizedOffer.kt
package com.onlycoffee.app.data.model

import com.google.gson.annotations.SerializedName

/**
 * Personalized offer model matching backend API
 */
data class PersonalizedOffer(
    val id: String,
    val title: String,
    val description: String,
    val offerType: OfferType,
    val discountValue: Double,
    val minPurchaseAmount: Double? = null,
    val maxDiscountAmount: Double? = null,
    val applicableMenuItemIds: List<String>? = null,
    val applicableCategories: List<String>? = null,
    val validFrom: String,
    val validUntil: String,
    val usageLimit: Int? = null,
    val usageCount: Int = 0,
    val source: OfferSource,
    val priority: Int = 0,
    val imageUrl: String? = null,
    val termsAndConditions: String? = null,
    val isActive: Boolean = true,
    val createdAt: String,
    val updatedAt: String
)

/**
 * Offer types
 */
enum class OfferType {
    @SerializedName("PERCENTAGE_OFF")
    PERCENTAGE_OFF,
    
    @SerializedName("FIXED_AMOUNT_OFF")
    FIXED_AMOUNT_OFF,
    
    @SerializedName("FREE_ITEM")
    FREE_ITEM,
    
    @SerializedName("BUY_ONE_GET_ONE")
    BUY_ONE_GET_ONE
}

/**
 * Offer sources
 */
enum class OfferSource {
    @SerializedName("ai_generated")
    AI_GENERATED,
    
    @SerializedName("manual")
    MANUAL,
    
    @SerializedName("geofence")
    GEOFENCE,
    
    @SerializedName("bandit")
    BANDIT,
    
    @SerializedName("trigger")
    TRIGGER
}

/**
 * Personalized offers response
 */
data class PersonalizedOffersResponse(
    val success: Boolean,
    val data: PersonalizedOffersData
)

data class PersonalizedOffersData(
    val offers: List<PersonalizedOffer>,
    val userSegment: String? = null,
    val churnRisk: Double? = null
)

/**
 * Extension functions for offer display
 */
fun PersonalizedOffer.getDisplayDiscount(): String {
    return when (offerType) {
        OfferType.PERCENTAGE_OFF -> "${discountValue.toInt()}% OFF"
        OfferType.FIXED_AMOUNT_OFF -> "$${discountValue.toInt()} OFF"
        OfferType.FREE_ITEM -> "FREE"
        OfferType.BUY_ONE_GET_ONE -> "BOGO"
    }
}

fun PersonalizedOffer.getSourceLabel(): String {
    return when (source) {
        OfferSource.AI_GENERATED -> "Personalized for You"
        OfferSource.MANUAL -> "Special Offer"
        OfferSource.GEOFENCE -> "Nearby Store Offer"
        OfferSource.BANDIT -> "Recommended"
        OfferSource.TRIGGER -> "Limited Time"
    }
}

fun PersonalizedOffer.isExpiringSoon(): Boolean {
    // Check if offer expires within 24 hours
    // TODO: Implement proper date parsing and comparison
    return false
}

fun PersonalizedOffer.canBeUsed(): Boolean {
    return isActive && (usageLimit == null || usageCount < usageLimit)
}

